### DISCUSSION TITLE
Module mfrc522 RFID

### DISCUSSION THREAD

**Discussion by salarizadi** [Moddable-OpenSource/moddable#1127] [Q&A]
Hello.

Is there a ready code to use the MFRC522 module for ESP32 ModdableSDK?

**Comment by phoddie** [comment]
Not that I'm aware of. NFC support would be great. I know @andycarle has given some thought to how that could fit into the ECMA-419 sensor / peripheral class patterns.

**Reply by salarizadi** [reply]
Surprise! Stay tuned for the module library for the Moddable SDK tomorrow night. I am writing esp codes

**Reply by salarizadi** [reply]
> Not that I'm aware of. NFC support would be great. I know @andycarle has given some thought to how that could fit into the ECMA-419 sensor / peripheral class patterns.

I use the esp_event_handler_register_with function from ESP-IDF. Unfortunately, after creating the task and scanning the RFID card, the Listener function is lost and I cannot access it. How can I have a function in an object to return RFID card information? How can I save the function and run it later?

Javascript : 
```
const $RC522 = new RC522({
    listener: function ( id ) {
        trace("Hello From listener :" + id)
    }
});
```

```
static void rc522_handler ( xsMachine *the, esp_event_base_t base, int32_t event_id, void* event_data ) {
    xsmcVars(1);

    rc522_event_data_t* data = (rc522_event_data_t*) event_data;

    switch(event_id) {
        case RC522_EVENT_TAG_SCANNED: {
                rc522_tag_t* tag = (rc522_tag_t*) data->ptr;

                if (xsmcHas(xsArg(0), xsID_listener)) {
                    xsCall0(xsArg(0), xsID_listener);
                } else {
                    xsTrace("Listener not found");
                }
            }
            break;
    }
}

void xs_rc522_init ( xsMachine *the ) {
    xsTrace("\nRC522 is running");

    rc522_config_t config = {
        .spi.host = VSPI_HOST,
        .spi.miso_gpio = 25,
        .spi.mosi_gpio = 23,
        .spi.sck_gpio = 19,
        .spi.sda_gpio = 22,
    };


    rc522_create(&config, &scanner);
    rc522_register_events(scanner, RC522_EVENT_ANY, rc522_handler, &the); // esp_event_handler_register_with
    rc522_start(scanner);
}
```

For example, for better understanding, I call an initial function to save the settings, then write a statement with while or FreeRtos. During this time, the javascript code changes, but how can I save and run this function for say another 1 minute or even longer?

**Reply by salarizadi** [reply]
Please reply so that I can complete the library as soon as possible🙏

**Reply by phoddie** [reply]
Sorry to be slow.

You are looking to invoke a JavaScript callback function from a native callback  in the Moddable SDK. That is done in many places in the Moddable SDK modules. The Wi-Fi module is one good place to look. The `WiFi.scan()` function accepts JavaScript callback as an argument, remembers it, and calls it during the scan.

One important detail: you **must** make all calls to XS for the virtual machine (VM) from the same native FreeRTOS task. If you do not, the VM state can be corrupted. We usually use the function `modMessagePostToMachine` to help do that. It lets you defer a native callback until it is safe to call with the specified virtual machine. In the case of `WiFi.scan()`, `modMessagePostToMachine` is called in the native ESP-IDF callback `doWiFiEvent` to defer calling the VM to the `reportScan` function.

**Reply by phoddie** [reply]
@salarizadi – was this information enough to get you unstuck? If not, I can try to put together a standalone example.

**Reply by salarizadi** [reply]
Unfortunately I haven't had time to check this yet, I will let you know soon.

If you like to explain other methods, that would be great. When I want to continue, if I have a problem I know how to fix it, so if you know something better, please let me know.

**Reply by phoddie** [reply]
> Unfortunately I haven't had time to check this yet, I will let you know soon.

No problem. Let me know how it goes. It is some work to build an example, so I would prefer to do that only if necessary.

**Reply by salarizadi** [reply]
@phoddie At first I was going to do this with C and ESP-IDF, but then I changed my mind. I decided to write a dedicated library for SPI ESP32 so that I can write the RF522 module using JavaScript. I am working on it, but I didn't have enough time because of other projects I have, but I will definitely let you know

**Reply by phoddie** [reply]
FYI - there is already a SPI module for ESP32. It is part of the ECMA-419 IO support. There some [examples here](https://github.com/Moddable-OpenSource/moddable/tree/public/examples/io/spi). There is also the [display driver for M5Core Ink](https://github.com/Moddable-OpenSource/moddable/blob/public/modules/drivers/gdew0154m09/gdew0154m09.js) that uses it and [display driver for M5 Paper](https://github.com/Moddable-OpenSource/moddable/blob/public/modules/drivers/it8951/it8951.js).

**Reply by salarizadi** [reply]
@phoddie It was interesting, I wish you would have provided a more complete document for SPI. Are you sure it meets all the expectations? Because in the document or from the screen or 2 simple lines without full guidance, the commands are said, I did not see a complete guidance on this matter.

**Reply by phoddie** [reply]
>  I wish you would have provided a more complete document for SPI

Here is a link to the [SPI section](https://419.ecma-international.org/#-10-io-classes-serial-peripheral-interface-spi) in the ECMA-419 Standard. You can read the [implementation](https://github.com/Moddable-OpenSource/moddable/blob/public/modules/io/spi/esp/spi.c).

> Are you sure it meets all the expectations? 

I don't know what you expect.

**Reply by salarizadi** [reply]
@phoddie Now that I have seen this information, maybe I can fix it. I will let you know in the next 48 hours. Thank you for your guidance

**Reply by salarizadi** [reply]
@phoddie 

Sorry for not being able to upload the code quickly enough. I ran into several problems. It is really a heavy project and I did not expect it to be so heavy.

How can I: send this array of hexadecimal houses to C and get the correct output from there?

JavaScript example:
```
const keys  = [
    0xFF, 0xFF, 0xFF,
    0xFF, 0xFF, 0xFF
];

class MFRC522 {
   _write ( block, data ) @ "xs_mfrc522_write";
    write ( block, data ) {
        if ( !Array.isArray(data) )
            trace("<error>MFRE522 write : data not is array");
        data = new Uint8Array(data).buffer;
        return this._write(block, data)
    }
}
```

Sample C:
```
MIFARE_Key key;
char *keys = xsmcToArrayBuffer(xsArg(2)); // Keys

for (int i = 0; i < sizeof(keys); i++)
        key.keyByte[i] = keys[i];

...

xsmcSetInteger(xsResult, PCD_Authenticate(MFRC522_SPI, cmd_k, block, &key, &uid));
```

If I solve this problem faster, the MFRC522 module will be uploaded in 24 hours

Thank you.

**Reply by phoddie** [reply]
If I understand correctly, your JavaScript `write` function accepts an Array. You want to treat that as an array of bytes in the native code. To do that, you can pass the Array through to the native code directly. I think that will be simpler and at least as efficient:

```js
class MFRC522 {
    write ( block, data ) @ "xs_mfrc522_write";
}
```

In C, you can access the array elements directly:

```c
xsmcVars(1);
xsmcGet(xsVar(0), xsArg(1), xsID_length);  // get the length property of argument 1
int length = xsmcToInteger(xsVar(0));  // get the value of the length property as an integer
int i;
for (i = 0; i < length; i++) {
   xsmcGetIndex(xsVar(0), xsArg(1), i); // get element i of array in argument 1
   uint8_t byte = (uint8_t)xsmcToInteger(xsVar(0));  // get value of element as integer
   key.keyByte[i] = byte;
}
```

**Reply by salarizadi** [reply]
@phoddie It worked correctly! Thankful

**Reply by salarizadi** [reply]
@phoddie 

What should I do if I want to add to a hexadecimal array and display it as hex? This example I wrote only shows the number:

```
Uid *cardID = &uid;

xsmcVars(1);

xsResult = xsmcNewArray(0);
for (int i = 0; i < cardID->size; i++) {
     xsVar(0) = xsInteger(cardID->uidByte[i]);
     xsmcSetIndex(xsResult, i, xsVar(0));
     printf("\n uid : %x", cardID->uidByte[i]);
}
```

Result :  ```[131,79,152,30]```

I want to show it like this : ```[83,4f,98,1e]```

And if I want to add this array to a key in the object, like:

```
trace(mfrc522.info())

{
  "id" : [83,4f,98,1e],
  "type": ...,
  ....
}
```

What should I do

note : These items are for returning RFID card information in the form of objects

**Reply by phoddie** [reply]
You can easily return an `ArrayBuffer` with a copy of a block of native memory. Please see the blog post "[Best Practices for Accessing Memory Buffers from Native Code](https://blog.moddable.com/blog/buffers/)" specifically in the section "Creating a Buffer".
